File: src-tauri/src/commands/recovery.rs
//! Reflog, destructive-operation previews, discard-all, and stranded index.lock recovery.

use super::{blocking, CommandError};
use crate::git;
use crate::git::types::{
    DeleteBranchPreview, DestructivePreview, DiscardAllPreview, ForcePushPreview, IndexLockStatus,
    ReflogEntry, ResetPreview,
};

#[tauri::command]
pub async fn list_reflog(
    path: String,
    limit: Option<usize>,
) -> Result<Vec<ReflogEntry>, CommandError> {
    // Clamp the caller-supplied limit: the UI requests 120, but the command
    // surface must not let a stray large value trigger an unbounded reflog walk.
    let limit = limit.unwrap_or(80).clamp(1, 500);
    blocking(move || git::write::recovery::reflog_entries(&path, limit)).await
}

#[tauri::command]
pub async fn preview_reset(
    path: String,
    target: String,
    mode: String,
    source: Option<String>,
) -> Result<ResetPreview, CommandError> {
    // `source` is the ref being reset; defaults to HEAD for current-branch resets.
    let source = source.unwrap_or_else(|| "HEAD".to_string());
    blocking(move || git::write::recovery::preview_reset(&path, &target, &mode, &source)).await
}

#[tauri::command]
pub async fn preview_discard_all(path: String) -> Result<DiscardAllPreview, CommandError> {
    blocking(move || git::write::discard_all::preview_discard_all(&path)).await
}

#[tauri::command]
pub async fn preview_delete_branch(
    path: String,
    branch: String,
) -> Result<DeleteBranchPreview, CommandError> {
    blocking(move || git::write::recovery::preview_delete_branch(&path, &branch)).await
}

#[tauri::command]
pub async fn preview_delete_remote_branch(
    path: String,
    remote: String,
    branch: String,
) -> Result<DestructivePreview, CommandError> {
    blocking(move || git::write::recovery::preview_delete_remote_branch(&path, &remote, &branch))
        .await
}

#[tauri::command]
pub async fn preview_force_push(
    path: String,
    branch: String,
) -> Result<ForcePushPreview, CommandError> {
    blocking(move || git::write::recovery::preview_force_push(&path, &branch)).await
}

#[tauri::command]
pub async fn discard_all(
    path: String,
    expected_state: String,
    expected_head_branch: Option<String>,
    expected_head_oid: Option<String>,
) -> Result<String, CommandError> {
    blocking(move || {
        git::write::discard_all::discard_all(
            &path,
            &expected_state,
            expected_head_branch.as_deref(),
            expected_head_oid.as_deref(),
        )
    })
    .await
}

/// Inspect `.git/index.lock` for the stranded-lock recovery toast (GL-335).
#[tauri::command]
pub async fn inspect_index_lock(path: String) -> Result<IndexLockStatus, CommandError> {
    blocking(move || git::write::index_lock::inspect_index_lock(&path)).await
}

/// Remove a stranded `.git/index.lock` only when the staleness gate passes.
#[tauri::command]
pub async fn remove_index_lock(path: String) -> Result<(), CommandError> {
    blocking(move || git::write::index_lock::remove_index_lock(&path)).await
}

File: src/lib/api/git/recovery.ts
// The reflog, the destructive-write previews that hand back the state lease a
// write must then present, the index-lock probe and its removal, and
// `discardAll` — the write those previews guard. Mirrors `commands/recovery.rs`.

import { invoke } from "@/lib/api/invoke";
import {
  deleteBranchPreviewSchema,
  destructivePreviewSchema,
  discardAllPreviewSchema,
  forcePushPreviewSchema,
  indexLockStatusSchema,
  reflogEntrySchema,
  resetPreviewSchema,
} from "@/lib/api/schemas";
import { parse } from "@/lib/api/validate";
import { z } from "zod";
import type {
  DeleteBranchPreview,
  DestructivePreview,
  DiscardAllPreview,
  ForcePushPreview,
  IndexLockStatus,
  ReflogEntry,
  ResetPreview,
} from "./types";

export const recoveryApi = {
  listReflog: async (path: string, limit?: number): Promise<ReflogEntry[]> =>
    parse(
      z.array(reflogEntrySchema),
      await invoke("list_reflog", { path, limit: limit ?? null }),
      "list_reflog",
    ),

  previewReset: async (
    path: string,
    target: string,
    mode: "soft" | "mixed" | "hard",
    // The ref being reset; omit for current-branch resets (defaults to HEAD).
    source?: string,
  ): Promise<ResetPreview> =>
    parse(
      resetPreviewSchema,
      await invoke("preview_reset", { path, target, mode, source: source ?? null }),
      "preview_reset",
    ),

  previewDiscardAll: async (path: string): Promise<DiscardAllPreview> =>
    parse(
      discardAllPreviewSchema,
      await invoke("preview_discard_all", { path }),
      "preview_discard_all",
    ),

  previewDeleteBranch: async (path: string, branch: string): Promise<DeleteBranchPreview> =>
    parse(
      deleteBranchPreviewSchema,
      await invoke("preview_delete_branch", { path, branch }),
      "preview_delete_branch",
    ),

  previewDeleteRemoteBranch: async (
    path: string,
    remote: string,
    branch: string,
  ): Promise<DestructivePreview> =>
    parse(
      destructivePreviewSchema,
      await invoke("preview_delete_remote_branch", { path, remote, branch }),
      "preview_delete_remote_branch",
    ),

  previewForcePush: async (path: string, branch: string): Promise<ForcePushPreview> =>
    parse(
      forcePushPreviewSchema,
      await invoke("preview_force_push", { path, branch }),
      "preview_force_push",
    ),

  /** Discard every uncommitted change: reset tracked files to HEAD and remove
   * untracked files/dirs. Irreversible. */
  discardAll: async (
    path: string,
    expectedState: string,
    expectedHeadBranch: string | null,
    expectedHeadOid: string | null,
  ) =>
    parse(
      z.string(),
      await invoke("discard_all", { path, expectedState, expectedHeadBranch, expectedHeadOid }),
      "discard_all",
    ),

  /** Inspect `.git/index.lock` for stranded-lock recovery (GL-335). */
  inspectIndexLock: async (path: string): Promise<IndexLockStatus> =>
    parse(
      indexLockStatusSchema,
      await invoke("inspect_index_lock", { path }),
      "inspect_index_lock",
    ),

  /** Remove a stranded `.git/index.lock` only when the staleness gate passes. */
  removeIndexLock: (path: string) => invoke<void>("remove_index_lock", { path }),
};
